### ISSUE TITLE
storaged: Memory leak in the face of cancellations of DML statements

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#12593]
### What version of Materialize are you using?

v0.26.1-dev (8156df7a3)

### How did you install Materialize?

Built from source

### What is the issue?

Memory is leaked in the face of cancellations. This seems to happen in at least the following two cases:
- hitting Ctrl+C manually during an UPDATE statement
- If the read timeout from #12427 kicks in

My guess is that clients simply disconnecting will also cause the same situation.

To reproduce:

Step 1: run the following test case :

```
drop table if exists t1;
create table t1 (f1 DOUBLE);
insert into t1 SELECT generate_series FROM generate_series(1,1000000);
update t1 set f1 = f1 + 1000000;
insert into t1 SELECT generate_series FROM generate_series(1,10000000);
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "1 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "2 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "3 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "4 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "5 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "6 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "7 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "8 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "9 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "10 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "11 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "12 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "13 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "14 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "15 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "16 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "17 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "18 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "19 s";
update t1 set f1 = f1 + 10000000;
-- SET statement_timeout = "20 s";
update t1 set f1 = f1 + 10000000;
```

Observe an expected see-saw pattern in memory consumption, all within a few GB

Step No 2. Run the test case above but with the `SET` statements enabled **OR** just hit Ctrl+C repeatedly every 0.5 seconds while the test is running.

Observe memory consumption go to 10-30 GB. Screenshot attached:
![Screenshot from 2022-05-20 10-18-19](https://user-images.githubusercontent.com/1279722/169569404-9d9afd70-5d6d-4aa9-ad31-2cbaf7180d34.png)

**Comment by philip-stoev** [comment]
Based on a comment by @antiguru , sending this to the Dataflow team.

**Comment by uce** [comment]
@antiguru Can you look into this and make the final call here?

**Comment by antiguru** [comment]
I have a hunch of what might be going on: We might push into an input where the corresponding stream has been dropped. This would currently cause Timely to buffer all data in a shared structure.

**Comment by antiguru** [comment]
So, it turns out this is a race condition between compute and storage:
1. Compute subscribes to a source
2. Compute receives a cancellation and unsubscribes from the source.
3. Storage catches up with its pending commands and announces the availability of a source and starts streaming data.
4. All data will remain in memory indefinitely because the tcp boundary assumes that there will be a client subscribing to the just-announced subscription.

To fix this, we could remember that a subscription is invalidated and not accept the announcement, but since this code will go away, I'm inclined to mark it fixed by #12216 once it mergres.
